### PR TITLE
Improve location reporting of trait placeholder error

### PR DESCRIPTION
**Pull request by compiler-errors**
Self-explanatory

### CODE REVIEW

File: compiler/rustc_borrowck/src/region_infer/mod.rs [hunk 1]
```diff
@@ -2055,14 +2055,13 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         let cause_code = path
             .iter()
             .find_map(|constraint| {
-                if let ConstraintCategory::Predicate(predicate_span) = constraint.category {
+                if let ConstraintCategory::Predicate(source_def_id, predicate_span) =
+                    constraint.category
+                {
                     // We currently do not store the `DefId` in the `ConstraintCategory`
                     // for performances reasons. The error reporting code used by NLL only
                     // uses the span, so this doesn't cause any problems at the moment.
```
Comment (fmease) [inline compiler/rustc_borrowck/src/region_infer/mod.rs:2063]:
```suggestion
```

or adjust accordingly?

Comment (compiler-errors) [reply compiler/rustc_borrowck/src/region_infer/mod.rs:2063]: LOL

File: tests/ui/coroutine/resume-arg-late-bound.stderr [hunk 2]
```diff
@@ -1,10 +1,13 @@
 error: implementation of `Coroutine` is not general enough
   --> $DIR/resume-arg-late-bound.rs:15:5
    |
+LL | fn test(a: impl for<'a> Coroutine<&'a mut bool>) {}
+   |                 ------------------------------- due to a where-clause on `test`...
```
Comment (cjgillot) [inline tests/ui/coroutine/resume-arg-late-bound.stderr:5]: The terminology may be a bit surprising here.

**Comment by compiler-errors** [comment]
> for performance purposes

@bors try @rust-timer queue

File: compiler/rustc_borrowck/src/type_check/canonical.rs [hunk 3]
```diff
@@ -101,13 +101,13 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
         &mut self,
         // Keep this parameter for now, in case we start using
         // it in `ConstraintCategory` at some point.
```
Comment (cjgillot) [inline compiler/rustc_borrowck/src/type_check/canonical.rs:103]:
```suggestion
```

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/placeholder_error.rs [hunk 4]
```diff
@@ -240,21 +240,17 @@ impl<'tcx> NiceRegionError<'_, 'tcx> {
     ) -> Diag<'tcx> {
         let span = cause.span();
 
-        let (leading_ellipsis, satisfy_span, where_span, dup_span, def_id) =
-            if let ObligationCauseCode::WhereClause(def_id, span)
-            | ObligationCauseCode::WhereClauseInExpr(def_id, span, ..) = *cause.code()
-                && def_id != CRATE_DEF_ID.to_def_id()
```
Comment (cjgillot) [inline compiler/rustc_infer/src/infer/error_reporting/nice_region_error/placeholder_error.rs:246]: Why is this check removed? `process_registered_region_obligations` still introduces a dummy crate def-id. Won't it cause issues?

**Comment by compiler-errors** [comment]
This is a nice improvement but I have no time for this anymore.

**Pull request closed without merging**